### ISSUE TITLE
Sphinx crashes when labelling directive '.. todolist::'

### ISSUE THREAD

**Issue by ghost** [sphinx-doc/sphinx#4063]
Subject: Sphinx crashes when labelling directive '.. todolist::'

The following minimal working document lets sphinx crash:

```
.. _just a reference:
.. todolist::
```

The extension sphinx.ext.todo is enabled (nothing else) and this was tested with sphinx 1.6.3 and the current git-master at time of writing (228fdb892af25f4b93f2760f2cd6497f2aabc0be dated 2017-09-17)
(Command used was 'make html' with the Makefile generated by sphinx-quickstart)
Output:

```
Running Sphinx v1.7+
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] index
Exception occurred:
  File "/usr/lib64/python2.7/site-packages/docutils/nodes.py", line 954, in replace_self
    'Losing "%s" attribute: %s' % (att, self[att])
AssertionError: Losing "ids" attribute: ['just-a-reference']
The full traceback has been saved in /tmp/sphinx-err-LDfgg6.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:21: html] Error 1
```

Log file (for current git head):
```
# Sphinx version: 1.7+
# Python version: 2.7.12 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.5
# Last messages:
#   
#   looking for now-outdated files...
#   none found
#   pickling environment...
#   done
#   checking consistency...
#   done
#   preparing documents...
#   done
#   writing output... [100%] index
# Loaded extensions:
#   sphinx.ext.todo (1.7+) from .../sphinx/sphinx/sphinx/ext/todo.py
#   alabaster (0.7.9) from /usr/lib64/python2.7/site-packages/alabaster/__init__.pyc
Traceback (most recent call last):
  File ".../sphinx/sphinx/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File ".../sphinx/sphinx/sphinx/application.py", line 330, in build
    self.builder.build_update()
  File ".../sphinx/sphinx/sphinx/builders/__init__.py", line 328, in build_update
    'out of date' % len(to_build))
  File ".../sphinx/sphinx/sphinx/builders/__init__.py", line 394, in build
    self.write(docnames, list(updated_docnames), method)
  File ".../sphinx/sphinx/sphinx/builders/__init__.py", line 431, in write
    self._write_serial(sorted(docnames))
  File ".../sphinx/sphinx/sphinx/builders/__init__.py", line 438, in _write_serial
    doctree = self.env.get_and_resolve_doctree(docname, self)
  File ".../sphinx/sphinx/sphinx/environment/__init__.py", line 883, in get_and_resolve_doctree
    self.apply_post_transforms(doctree, docname)
  File ".../sphinx/sphinx/sphinx/environment/__init__.py", line 935, in apply_post_transforms
    self.app.emit('doctree-resolved', doctree, docname)
  File ".../sphinx/sphinx/sphinx/application.py", line 436, in emit
    return self.events.emit(event, self, *args)
  File ".../sphinx/sphinx/sphinx/events.py", line 79, in emit
    results.append(callback(*args))
  File ".../sphinx/sphinx/sphinx/ext/todo.py", line 182, in process_todo_nodes
    node.replace_self(content)
  File "/usr/lib64/python2.7/site-packages/docutils/nodes.py", line 954, in replace_self
    'Losing "%s" attribute: %s' % (att, self[att])
AssertionError: Losing "ids" attribute: ['just-a-reference']
```

**Comment by shibukawa** [comment]
This stacktrace is shown when ``todo_include_todos = False``.
But otherwise, the following error happens:
```Traceback (most recent call last):
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/application.py", line 339, in build
    self.builder.build_update()
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/builders/__init__.py", line 331, in build_update
    'out of date' % len(to_build))
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/builders/__init__.py", line 397, in build
    self.write(docnames, list(updated_docnames), method)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/builders/__init__.py", line 434, in write
    self._write_serial(sorted(docnames))
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/builders/__init__.py", line 441, in _write_serial
    doctree = self.env.get_and_resolve_doctree(docname, self)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/environment/__init__.py", line 911, in get_and_resolve_doctree
    self.apply_post_transforms(doctree, docname)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/environment/__init__.py", line 963, in apply_post_transforms
    self.app.emit('doctree-resolved', doctree, docname)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/application.py", line 489, in emit
    return self.events.emit(event, self, *args)
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/events.py", line 79, in emit
    results.append(callback(*args))
  File "/Users/shibu/dev/sphinx/sphinx/sphinx/ext/todo.py", line 184, in process_todo_nodes
    node.replace_self(content)
  File "/Users/shibu/dev/sphinx/lib/python3.6/site-packages/docutils/nodes.py", line 948, in replace_self
    update.update_basic_atts(self)
  File "/Users/shibu/dev/sphinx/lib/python3.6/site-packages/docutils/nodes.py", line 705, in update_basic_atts
    self.append_attr_list(att, dict_.get(att, []))
  File "/Users/shibu/dev/sphinx/lib/python3.6/site-packages/docutils/nodes.py", line 717, in append_attr_list
    if not value in self[attr]:
  File "/Users/shibu/dev/sphinx/lib/python3.6/site-packages/docutils/nodes.py", line 567, in __getitem__
    return self.attributes[key]
KeyError: 'ids'```

**Comment by tk0miya** [comment]
Fixed by #4144
